DuckDuckGoTests/BookmarkEntityTests.swift

```diff
@@ -0,0 +1,217 @@
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import XCTest
+import Persistence
+import CoreData
+import Bookmarks
+import DuckDuckGo
+
+class BookmarkEntityTests: XCTestCase {
+
+    var db: CoreDataDatabase!
+    var context: NSManagedObjectContext!
+    var root: BookmarkEntity!
+    var favorites: BookmarkEntity!
+
+    override func setUpWithError() throws {
+        try super.setUpWithError()
+        let model = CoreDataDatabase.loadModel(from: Bookmarks.bundle, named: "BookmarksModel")!
+
+        db = CoreDataDatabase(name: "Test", containerLocation: tempDBDir(), model: model)
+        db.loadStore()
+
+        context = db.makeContext(concurrencyType: .mainQueueConcurrencyType)
+        BookmarkUtils.prepareFoldersStructure(in: context)
+
+        try context.save()
+
+        root = BookmarkUtils.fetchRootFolder(context)
+        XCTAssertNotNil(root)
+
+        favorites = BookmarkUtils.fetchFavoritesFolder(context)
+        XCTAssertNotNil(favorites)
+    }
+
+    override func tearDownWithError() throws {
+        try super.tearDownWithError()
+
+        root = nil
+        favorites = nil
+        context = nil
+
+        try db.tearDown(deleteStores: true)
+    }
+
+    func testWhenBookmarkIsValidThenItSaves() throws {
+        _ = BookmarkEntity.makeBookmark(title: "t",
+                                        url: "u",
+                                        parent: root,
+                                        context: context)
+
+        try context.save()
+    }
+
+    func testWhenFavoriteIsValidThenItSaves() throws {
+        let favorite = BookmarkEntity.makeBookmark(title: "t",
+                                                   url: "u",
+                                                   parent: root,
+                                                   context: context)
+        favorite.addToFavorites(favoritesRoot: favorites)
+
+        try context.save()
+    }
+
+    func testWhenFolderIsValidThenItSaves() throws {
+        _ = BookmarkEntity.makeFolder(title: "f", parent: root, context: context)
+
+        try context.save()
+    }
+
+    func testWhenBookmarkIsMissingURLThenValidationFails() {
+        let bookmark = BookmarkEntity.makeBookmark(title: "t",
+                                                   url: "u",
+                                                   parent: root,
+                                                   context: context)
+
+        bookmark.url = nil
+
+        do {
+            try context.save()
+            XCTFail("Save should fail")
```

Comment:
Just a remark that these do/catch statements can be rewritten in a more concise form:
```swift
XCTAssertThrowsError(try context.save())
```